File: src/main/java/compiler/block/Block.java
package compiler.block;

import java.util.ArrayList;
import java.util.Collections;

/**
 * Represents a block of code.
 */
public abstract class Block {

    protected static int TOTAL_BLOCKS = 0;
    protected ArrayList<Block> subBlocks;

    // true ifs block can store sub blocks
    protected boolean container = false;
    // Block ID used for reference when storing in local variable table
    protected int id;
    private Block superBlock;
    private boolean variable = false;


    public Block(Block superBlock, boolean container, boolean variable) {
        this.setSuperBlock(superBlock);
        this.subBlocks = new ArrayList<>();
        this.container = container;
        id = TOTAL_BLOCKS++;

    }

    public Block getSuperBlock() {
        return superBlock;
    }

    public void setSuperBlock(Block superBlock) {
        this.superBlock = superBlock;
    }

    public ArrayList<Block> getBlockTree() {
        ArrayList<Block> blocks = new ArrayList<Block>();

        Block block = this;

        do {
            blocks.add(block);
            block = block.getSuperBlock();
        } while (block != null);

        Collections.reverse(blocks);

        return blocks;
    }

    public Block[] getSubBlocks() {
        return subBlocks.toArray(new Block[subBlocks.size()]);
    }

    public void addBlock(Block block) {
        subBlocks.add(block);
    }

    // Removes a block
    public void removeBlock(Block block) {
        subBlocks.remove(block);
    }

    // Called before looping through blocks to generate code. Allows for method to be called when all blocks are loaded
    public abstract void init();

    public abstract String getName();

    public abstract String getValue();

    public abstract String getType();

    public abstract String getOpeningCode();

    public abstract String getBodyCode();

    public abstract String getClosingCode();

    public int getId() {
        return id;
    }

    public void setId(int id) {
        this.id = id;
    }

    public boolean isContainer() {
        return container;
    }

    public void setContainer(boolean container) {
        this.container = container;
    }

    public boolean isVariable() {
        return variable;
    }

    public void setVariable(boolean variable) {
        this.variable = variable;
    }
}

File: src/main/scala/compiler/block/Block.scala
package compiler.block

import java.util
import java.util.ArrayList
import java.util.Collections

import asm.ASMGenerator

/**
  *
  */
object Block {
  var TOTAL_BLOCKS: Int = 50
}

/**
  * Represents a block of code.
  *
  * @param superBlockInitTest The superBlock of this block
  * @param containerInit      Whether the block contains other blocks
  * @param variableInit       Whether the block is a variable
  */
abstract class Block(var superBlockInitTest: Block, val containerInit: Boolean, val variableInit: Boolean) {

  Block.TOTAL_BLOCKS += 1

  private val _subBlocks: java.util.ArrayList[Block] = new ArrayList[Block]
  private val _asm: ASMGenerator = new ASMGenerator;
  private var _id: Integer = Block.TOTAL_BLOCKS
  private var _superBlock: Block = superBlockInitTest
  private var _container: Boolean = containerInit
  private var _variable: Boolean = variableInit

  /* id GET and SET */
  def id = _id

  def id_=(value: Integer) = _id = value

  /* superBlock GET and SET */
  def superBlock = _superBlock

  def superBlock_=(value: Block) = _superBlock = value

  /* subBlocks GET */
  def subBlocks = _subBlocks.toArray(new Array[Block](_subBlocks.size()))

  def addBlock_=(value: Block) = _subBlocks.add(value)

  def removeBlock_=(value: Block) = _subBlocks.remove(value)

  // Called after file is parsed
  def init()

  /* Symbol table information */
  def getName: String

  def getValue: String

  def getType: String

  /* Bytecode for the opening and closing of the block */
  def getOpeningCode: String

  def getClosingCode: String

  /* Whether the block contains other blocks */
  def isContainer: Boolean = _container

  def container_=(value: Boolean) = _container = value

  /* Whether the block is  a variable */
  def isVariable: Boolean = _variable

  def variable_=(value: Boolean) = _variable = value

  def error(): Boolean = false

  def asm: ASMGenerator = _asm
}